File: 백준/Gold/17298. 오큰수/오큰수.py
n = int(input())
arr = list(map(int, input().split()))[::-1]
ans = [-1] * n
stack = []
# Ai의 오큰수는 오른쪽에 있으면서
# Ai보다 큰 수 중에서 가장 왼쪽에 있는 수를 의미한다.

for i in range(n):
    now = arr[i]
    # 큐에 현재보다 왼쪽에 있는 애들 들어있음
    while stack:
        if stack[-1] <= now:
            stack.pop()
        else:
            # 1부터 시교
            ans[i] = stack[-1]
            break
    stack.append(arr[i])
print(*reversed(ans))

File: 백준/Gold/17298. 오큰수/오큰수.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(br.readLine());

    // split 대신 StringTokenizer 사용 (빠름)
    StringTokenizer st = new StringTokenizer(br.readLine());
    String[] input = new String[N];
    for (int i = 0; i < N; i++) input[i] = st.nextToken();

    int[] ans = new int[N];
    Deque<Integer> stack = new ArrayDeque<>();

    // === 아래 로직은 기존 알고리즘 그대로 ===
    for (int i = N - 1; i >= 0; i--) {
      int num = Integer.parseInt(input[i]);
      boolean flag = false;
      while (!flag) {
        if (!stack.isEmpty() && stack.peek() > num) {
          ans[i] = stack.peek();
          stack.addFirst(num);
          flag = true;
        } else if (!stack.isEmpty() && stack.peek() <= num) {
          stack.pop();
        } else {
          stack.addFirst(num);
          ans[i] = -1;
          flag = true;
        }
      }
    }

    // 루프 내 print 대신 StringBuilder로 한 번에 출력 (빠름)
    StringBuilder sb = new StringBuilder();
    for (int x : ans) sb.append(x).append(' ');
    System.out.print(sb);
  }
}
